salt/top.sls.in

```diff
@@ -1,22 +1,51 @@
-metalk8s-@@VERSION:
-  'metalk8s:node:version:@@VERSION':
-    - match: pillar
-    - metalk8s.roles.minion
-  'I@metalk8s:node:version:@@VERSION and I@metalk8s:node:roles:bootstrap':
-    - match: compound
-    - metalk8s.roles.bootstrap
-    - metalk8s.roles.salt-master
-    - metalk8s.roles.registry
-    - metalk8s.roles.repository
-  'I@metalk8s:node:version:@@VERSION and I@metalk8s:node:roles:ca':
-    - match: compound
-    - metalk8s.roles.ca
-  'I@metalk8s:node:version:@@VERSION and I@metalk8s:node:roles:etcd':
-    - match: compound
-    - metalk8s.roles.etcd
-  'I@metalk8s:node:version:@@VERSION and I@metalk8s:node:roles:master':
-    - match: compound
-    - metalk8s.roles.master
-  'I@metalk8s:node:version:@@VERSION and I@metalk8s:node:roles:node':
-    - match: compound
-    - metalk8s.roles.node
+{%- set version = "@@VERSION" %}
+
+{%- set version_match = "metalk8s:nodes:" ~ grains['id'] ~ ":version:" ~ version %}
+
+{% macro role_match(name) -%}
+metalk8s:nodes:{{ grains['id'] }}:roles:{{ name }}
+{%- endmacro %}
+
+{% macro role(name='', states=[]) -%}
+{%- if name %}
+  'I@{{ version_match }} and I@{{ role_match(name) }}':
+{%- else %}
+  'I@{{ version_match }}':
+{%- endif %}
+    {{ ([{ "match": "compound" }] + states) | yaml }}
+{%- endmacro %}
+
+metalk8s-{{ version }}:
+  # 'Default' role applicable to all nodes
+  {{ role(states=[
+       'metalk8s.roles.minion',
+     ])
+  }}
+
+  {{ role('bootstrap', [
+       'metalk8s.roles.bootstrap',
+       'metalk8s.roles.salt-master',
+       'metalk8s.roles.registry',
+       'metalk8s.roles.repository',
```

Comment:
Why include all this state ? Shouldn't be part directly of the `roles.bootstrap` salt states.

As we basically need registry, repository and salt-master during the bootstrap process

Reply:
Primarily, because I didn't want to make any 'observable' changes in this PR. Indeed, these are 'wrong' in a sense, the intent is to have role-specific states here only. But it's up to the Salt state refactoring PR to do so ;-)

Reply:
Oh ok I will say to the guys working on Salt state refactoring to do this :p